Grafana panel legend format strings missing variables

### Bug Description

The legend format strings in the charm source code of the form  
```
"legendFormat": "{{project_name}} | {{instance_name}} | {{uuid}}"
```
Take the form  
```
"legendFormat": " |  |  "
```
upon proxying the dashboards through cos-proxy.
This issue is not observed upon relating with the legacy grafana charm.

### To Reproduce

```
juju deploy --channel edge cos-proxy
juju deploy --channel edge prometheus-libvirt-exporter ple
juju relate ple nova-compute
juju relate ple cos-proxy:prometheus-target
juju relate ple cos-proxy:dashboards
```

### Environment

```
miles@mertkirpici-bastion:~$ juju status -m cos
Model  Controller  Cloud/Region  Version  SLA          Timestamp
cos    vader       k/localhost   2.9.38   unsupported  14:18:44Z

App                           Version  Status  Scale  Charm                         Channel  Rev  Address         Exposed  Message
alertmanager                  0.25.0   active      1  alertmanager-k8s              edge      64  10.152.183.66   no       
catalogue                              active      1  catalogue-k8s                 edge      14  10.152.183.140  no       
grafana                       9.2.1    active      1  grafana-k8s                   edge      76  10.152.183.231  no       
loki                          2.7.4    active      1  loki-k8s                      edge      80  10.152.183.246  no       
prometheus                    2.42.0   active      1  prometheus-k8s                edge     119  10.152.183.71   no       
prometheus-scrape-config-k8s  n/a      active      1  prometheus-scrape-config-k8s  edge      39  10.152.183.132  no       
traefik                       2.9.6    active      1  traefik-k8s                   edge     124  10.5.100.100    no 
```
```
miles@mertkirpici-bastion:~$ juju status -m openstack
Model      Controller  Cloud/Region             Version  SLA          Timestamp
openstack  vader       serverstack/serverstack  2.9.38   unsupported  14:19:16Z

SAAS                          Status  Store  URL
grafana-dashboards            active  vader  admin/cos.grafana-dashboards
prometheus-scrape-config-k8s  active  vader  admin/cos.prometheus-scrape-config-k8s

App                                 Version  Status   Scale  Charm                          Channel       Rev  Exposed  Message
cinder                              20.1.0   active       1  cinder                         yoga/stable   603  no       Unit is ready
cinder-mysql-router                 8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
cos-proxy                           n/a      active       1  cos-proxy                      edge           36  no       
glance                              24.1.0   active       1  glance                         yoga/stable   562  no       Unit is ready
glance-mysql-router                 8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
grafana-agent                                blocked      7  grafana-agent                  edge            5  no       Missing relation: 'logging-consumer'
horizon                             22.1.0   active       1  openstack-dashboard            yoga/stable   566  no       Unit is ready
keystone                            21.0.0   active       1  keystone                       yoga/stable   595  no       Application Ready
keystone-mysql-router               8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
mysql                               8.0.32   active       3  mysql-innodb-cluster           8.0/stable     43  no       Unit is ready: Mode: R/W, Cluster is ONLINE and can tolerate up to ONE failure.
neutron-api                         20.2.0   active       1  neutron-api                    yoga/stable   547  no       Unit is ready
neutron-api-mysql-router            8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
neutron-api-plugin-ovn              20.2.0   active       1  neutron-api-plugin-ovn         yoga/stable    29  no       Unit is ready
nova-cloud-controller               25.1.0   active       1  nova-cloud-controller          yoga/stable   634  no       Unit is ready
nova-cloud-controller-mysql-router  8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
nova-compute                        25.1.0   active       6  nova-compute                   yoga/stable   650  no       Unit is ready
nrpe                                         active       1  nrpe                           edge           99  no       Ready
ovn-central                         22.03.0  active       3  ovn-central                    22.03/stable   57  no       Unit is ready (leader: ovnnb_db)
ovn-chassis                         22.03.0  active       6  ovn-chassis                    22.03/stable  118  no       Unit is ready
placement                           7.0.0    active       1  placement                      yoga/stable    74  no       Unit is ready
placement-mysql-router              8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
ple                                          active       6  prometheus-libvirt-exporter    edge           16  no       Ready
poe                                          active       1  prometheus-openstack-exporter                 18  no       Ready
rabbitmq-server                     3.9.13   active       1  rabbitmq-server                3.9/stable    165  no       Unit is ready
vault                               1.7.9    active       1  vault                          1.7/stable     95  no       Unit is ready (active: true, mlock: enabled)
vault-mysql-router                  8.0.32   active       1  mysql-router                   8.0/edge       61  no       Unit is ready
```

### Relevant log output

```shell
ubuntu@juju-43758e-openstack-39:/var/log/juju$ cat unit-cos-proxy-5.log 
2023-04-10 12:02:25 INFO juju unit_agent.go:289 Starting unit workers for "cos-proxy/5"
2023-04-10 12:02:25 INFO juju.worker.apicaller connect.go:163 [656710] "unit-cos-proxy-5" successfully connected to "10.5.3.181:17070"
2023-04-10 12:02:25 INFO juju.worker.apicaller connect.go:260 [656710] password changed for "unit-cos-proxy-5"
2023-04-10 12:02:25 INFO juju.worker.apicaller connect.go:163 [656710] "unit-cos-proxy-5" successfully connected to "10.5.3.181:17070"
2023-04-10 12:02:26 INFO juju.worker.migrationminion worker.go:142 migration phase is now: NONE
2023-04-10 12:02:26 INFO juju.worker.logger logger.go:120 logger worker started
2023-04-10 12:02:26 INFO juju.worker.upgrader upgrader.go:216 no waiter, upgrader is done
2023-04-10 12:02:26 ERROR juju.worker.meterstatus runner.go:91 error running "meter-status-changed": charm missing from disk
2023-04-10 12:02:26 INFO juju.worker.uniter uniter.go:326 unit "cos-proxy/5" started
2023-04-10 12:02:26 INFO juju.worker.uniter uniter.go:631 resuming charm install
2023-04-10 12:02:26 INFO juju.worker.uniter.charm bundles.go:78 downloading ch:amd64/jammy/cos-proxy-36 from API server
2023-04-10 12:02:33 INFO juju.worker.uniter uniter.go:344 hooks are retried true
2023-04-10 12:02:33 INFO juju.worker.uniter.storage resolver.go:127 initial storage attachments ready
2023-04-10 12:02:33 INFO juju.worker.uniter resolver.go:149 found queued "install" hook
2023-04-10 12:02:34 INFO unit.cos-proxy/5.juju-log server.go:316 Running legacy hooks/install.
2023-04-10 12:02:39 INFO juju.worker.uniter.operation runhook.go:146 ran "install" hook (via hook dispatching script: dispatch)
2023-04-10 12:02:39 INFO juju.worker.uniter resolver.go:149 found queued "leader-elected" hook
2023-04-10 12:02:39 INFO juju.worker.uniter.operation runhook.go:146 ran "leader-elected" hook (via hook dispatching script: dispatch)
2023-04-10 12:02:40 INFO juju.worker.uniter.operation runhook.go:146 ran "config-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:02:40 INFO juju.worker.uniter resolver.go:149 found queued "start" hook
2023-04-10 12:02:40 INFO unit.cos-proxy/5.juju-log server.go:316 Running legacy hooks/start.
2023-04-10 12:02:41 INFO juju.worker.uniter.operation runhook.go:146 ran "start" hook (via hook dispatching script: dispatch)
2023-04-10 12:08:09 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:09:59 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-prometheus-scrape-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:00 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-prometheus-scrape-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:01 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-prometheus-scrape-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:11 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:12 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:13 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:13 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:14 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:15 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:16 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:16 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:17 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:18 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:19 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:20 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:21 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:22 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:23 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:24 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:25 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:26 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:26 INFO juju.worker.uniter.operation runhook.go:146 ran "prometheus-target-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:29 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:30 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:31 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:10:32 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:11:17 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-grafana-dashboard-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 12:11:18 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-grafana-dashboard-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 12:11:18 INFO juju.worker.uniter.operation runhook.go:146 ran "downstream-grafana-dashboard-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 12:12:48 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:17:26 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:21:45 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:27:30 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:32:11 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:37:58 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:42:44 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:48:34 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:53:09 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 12:58:54 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:03:14 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:07:31 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:13:27 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:17:30 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:22:13 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:26:26 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:31:57 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:37:56 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:42:48 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:00 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-created" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:01 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:01 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: NOTHING!
2023-04-10 13:45:01 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: Could not find dashboard data after a relation change for <ops.model.Application ple>
2023-04-10 13:45:01 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:02 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:02 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: NOTHING!
2023-04-10 13:45:02 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: Could not find dashboard data after a relation change for <ops.model.Application ple>
2023-04-10 13:45:03 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:03 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:04 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: NOTHING!
2023-04-10 13:45:04 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: Could not find dashboard data after a relation change for <ops.model.Application ple>
2023-04-10 13:45:04 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:05 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:06 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:06 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:07 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: NOTHING!
2023-04-10 13:45:07 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: Could not find dashboard data after a relation change for <ops.model.Application ple>
2023-04-10 13:45:07 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:08 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-joined" hook (via hook dispatching script: dispatch)
2023-04-10 13:45:08 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: NOTHING!
2023-04-10 13:45:08 WARNING unit.cos-proxy/5.juju-log server.go:316 dashboards:164: Could not find dashboard data after a relation change for <ops.model.Application ple>
2023-04-10 13:45:09 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 13:48:05 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:53:20 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 13:58:17 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 14:02:34 INFO juju.worker.uniter.operation runhook.go:146 ran "dashboards-relation-changed" hook (via hook dispatching script: dispatch)
2023-04-10 14:04:09 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 14:08:43 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 14:14:03 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 14:19:37 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
2023-04-10 14:24:18 INFO juju.worker.uniter.operation runhook.go:146 ran "update-status" hook (via hook dispatching script: dispatch)
```


### Additional context

![image](https://user-images.githubusercontent.com/68030795/230920334-b502044f-aeee-434c-bf25-4934285cab79.png)
![image](https://user-images.githubusercontent.com/68030795/230920528-58994a09-2aee-4c9e-911b-f5655f61788f.png)
https://git.launchpad.net/charm-prometheus-libvirt-exporter/tree/src/files/grafana-dashboards/libvirt.json#n739

--- a/lib/charms/grafana_k8s/v0/grafana_dashboard.py
+++ b/lib/charms/grafana_k8s/v0/grafana_dashboard.py
@@ -219,7 +219,7 @@ LIBAPI = 0
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
 
-LIBPATCH = 28
+LIBPATCH = 29
 
 logger = logging.getLogger(__name__)
 
@@ -1417,11 +1417,6 @@ class GrafanaDashboardConsumer(Object):
         # The only piece of data needed on this side of the relations is "templates"
         templates = data.pop("templates")
 
-        # Import only if a charmed operator uses the consumer, we don't impose these
-        # dependencies on the client
-        from jinja2 import Template
-        from jinja2.exceptions import TemplateSyntaxError
-
         # The dashboards are WAY too big since this ultimately calls out to Juju to
         # set the relation data, and it overflows the maximum argument length for
         # subprocess, so we have to use b64, annoyingly.
@@ -1434,14 +1429,12 @@ class GrafanaDashboardConsumer(Object):
         relation_has_invalid_dashboards = False
 
         for _, (fname, template) in enumerate(templates.items()):
-            decoded_content = None
             content = None
             error = None
             topology = template.get("juju_topology", {})
             try:
-                decoded_content = _decode_dashboard_content(template["content"])
+                content = _decode_dashboard_content(template["content"])
                 inject_dropdowns = template.get("inject_dropdowns", True)
-                content = Template(decoded_content).render()
                 content = self._manage_dashboard_uid(content, template)
                 content = _convert_dashboard_fields(content, inject_dropdowns)
 
@@ -1456,9 +1449,6 @@ class GrafanaDashboardConsumer(Object):
                 error = str(e.msg)
                 logger.warning("Invalid JSON in Grafana dashboard: {}".format(fname))
                 continue
-            except TemplateSyntaxError as e:
-                error = str(e)
-                relation_has_invalid_dashboards = True
 
             # Prepend the relation name and ID to the dashboard ID to avoid clashes with
             # multiple relations with apps from the same charm, or having dashboards with
@@ -1830,10 +1820,10 @@ class GrafanaDashboardAggregator(Object):
             # Replace old piechart panels
             dash = re.sub(r'"type": "grafana-piechart-panel"', '"type": "piechart"', dash)
 
-            from jinja2 import Template
+            from jinja2 import DebugUndefined, Template
 
             content = _encode_dashboard_content(
-                Template(dash).render(host=r"$host", datasource=r"${prometheusds}")  # type: ignore
+                Template(dash, undefined=DebugUndefined).render(datasource=r"${prometheusds}")  # type: ignore
             )
             id = "prog:{}".format(content[-24:-16])
 

--- a/tests/unit/test_charm.py
+++ b/tests/unit/test_charm.py
@@ -75,7 +75,10 @@ DASHBOARD_DUMMY_DATA_1 = {
                             {"datasource": "Juju data"},
                         ],
                     },
-                    "panels": {"data": "some_data_to_hash_across"},
+                    "panels": {
+                        "data": "some_data_to_hash_across",
+                        "legendFormat": "{{ host }} | {{ instance_name }}",
+                    },
                 },
             },
         }
@@ -86,7 +89,7 @@ DUMMY_FIXED_1 = {
     "charm": "dashboard-app-1",
     "content": '{"__inputs": [], "templating": {"list": [{"datasource": '
     '"${prometheusds}"}]}, "panels": {"data": '
-    '"some_data_to_hash_across"}}',
+    '"some_data_to_hash_across", "legendFormat": "{{ host }} | {{ instance_name }}"}}',
     "inject_dropdowns": True,
     "juju_topology": {
         "application": "dashboard-app-1",
@@ -106,7 +109,10 @@ DASHBOARD_DUMMY_DATA_2 = {
                             {"name": "host"},
                         ],
                     },
-                    "panels": {"data": "different_enough_to_rehash"},
+                    "panels": {
+                        "legendFormat": "{{ host }} | {{ instance_name }}",
+                        "data": "different_enough_to_rehash",
+                    },
                 },
             },
         }
@@ -116,7 +122,8 @@ DASHBOARD_DUMMY_DATA_2 = {
 DUMMY_FIXED_2 = {
     "charm": "dashboard-app-2",
     "content": '{"templating": {"list": [{"name": "host"}]}, '
-    '"panels": {"data": "different_enough_to_rehash"}}',
+    '"panels": {"legendFormat": "{{ host }} | {{ instance_name }}", '
+    '"data": "different_enough_to_rehash"}}',
     "inject_dropdowns": True,
     "juju_topology": {
         "application": "dashboard-app-2",
@@ -855,5 +862,5 @@ class COSProxyCharmTest(unittest.TestCase):
         )
         dashboards = json.loads(grafana_rel_data.get("dashboards", "{}"))
         self.assertEqual(len(dashboards["templates"]), 2)
-        self.assertEqual(dashboards["templates"]["prog:e_data_t"], DUMMY_FIXED_1)
+        self.assertEqual(dashboards["templates"]["prog:| {{ ins"], DUMMY_FIXED_1)
         self.assertEqual(dashboards["templates"]["prog:rent_eno"], DUMMY_FIXED_2)

